server/src/main/java/org/elasticsearch/health/node/selection/HealthNodeSelectorTaskExecutor.java

```diff
@@ -0,0 +1,129 @@
+import org.elasticsearch.persistent.AllocatedPersistentTask;
+import org.elasticsearch.persistent.PersistentTaskParams;
+import org.elasticsearch.persistent.PersistentTaskState;
+import org.elasticsearch.persistent.PersistentTasksCustomMetadata;
+import org.elasticsearch.persistent.PersistentTasksExecutor;
+import org.elasticsearch.persistent.PersistentTasksService;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.RemoteTransportException;
+import org.elasticsearch.xcontent.NamedXContentRegistry;
+import org.elasticsearch.xcontent.ParseField;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.elasticsearch.health.node.selection.HealthNodeSelector.TASK_NAME;
+
+/**
+ * Persistent task executor that is managing the {@link HealthNodeSelector}.
+ */
+public final class HealthNodeSelectorTaskExecutor extends PersistentTasksExecutor<HealthNodeSelectorTaskParams> {
+
+    private static final Logger logger = LogManager.getLogger(HealthNodeSelector.class);
+
+    private final ClusterService clusterService;
+    private final PersistentTasksService persistentTasksService;
+    private final AtomicReference<HealthNodeSelector> currentTask = new AtomicReference<>();
+
+    public HealthNodeSelectorTaskExecutor(Client client, ClusterService clusterService, ThreadPool threadPool) {
+        super(TASK_NAME, ThreadPool.Names.MANAGEMENT);
+        this.clusterService = clusterService;
+        persistentTasksService = new PersistentTasksService(clusterService, threadPool, client);
+    }
+
+    @Override
+    protected void nodeOperation(AllocatedPersistentTask task, HealthNodeSelectorTaskParams params, PersistentTaskState state) {
+        HealthNodeSelector healthNodeSelector = (HealthNodeSelector) task;
+        currentTask.set(healthNodeSelector);
+        String nodeId = clusterService.localNode().getId();
+        logger.info("Node [" + nodeId + "] is selected as the current health node.");
+    }
+
+    @Override
+    protected HealthNodeSelector createTask(
+        long id,
+        String type,
+        String action,
+        TaskId parentTaskId,
+        PersistentTasksCustomMetadata.PersistentTask<HealthNodeSelectorTaskParams> taskInProgress,
+        Map<String, String> headers
+    ) {
+        return new HealthNodeSelector(id, type, action, getDescription(taskInProgress), parentTaskId, headers);
+    }
+
+    /**
+     * Returns the node id from the eligible health nodes
+     */
+    @Override
+    public PersistentTasksCustomMetadata.Assignment getAssignment(
+        HealthNodeSelectorTaskParams params,
+        Collection<DiscoveryNode> candidateNodes,
+        ClusterState clusterState
+    ) {
+        DiscoveryNode discoveryNode = selectLeastLoadedNode(clusterState, candidateNodes, node -> node.isMasterNode() == false);
```

Comment:
Doesn't this predicate mean that for single-node clusters (like someone starting ES locally), then we won't be able to select a node because it will be the master node with no other non-master nodes available?

Reply:
Very good point, that is correct. Hm.... In that case, when we have a single node cluster we should make the master node the health node. It's not acceptable to have the health API not working, and I would like to not have different flows even if that means that this will not be very efficient. What do you think?

Reply:
If we are in a single-node cluster, then we should probably allow the master node to be the health node. I'm thinking though, about what is going to happen when a cluster is started up quickly starting with the master node first—it seems like it'd be possible for the task to be assigned to the master node (because it's the only node in the cluster), but then have other nodes added later. Perhaps we should handle this situation by cancelling the task if it exists on the master node once the cluster reaches >= 2 nodes in total?

Reply:
Should we have the predicate be `nodeRole.canHoldData == true` ?

Reply:
I see, this way we do not exclude master nodes if they are both master nodes if they could hold data. It becomes a matter of chance, we were pretty adamant that we didn't want the master node to get extra load if possible so I am worrying that this is not good enough. Do you think checking the number of nodes is too complex?

Reply:
I think most very small deployments don't have dedicated master nodes, at which point I think it's a generally acceptable consequence that the master node does other things too. 
For medium-large enough deployments this will guarantee the master node will not host the health task.

IMO this is a good enough invariant.

Reply:
Okay, I will change the code

Reply:
Fixed